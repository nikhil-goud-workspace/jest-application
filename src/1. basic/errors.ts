export function failIfNegative(n: number) {
    if (n < 0)
        throw new Error('Negative number!');
    return n;
}
