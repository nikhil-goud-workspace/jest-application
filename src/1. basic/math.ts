export function add(a: number, b: number) {
    return a + b;
}
export function getObject() {
    return { a: 1, b: 2 };
}
export function isPositive(n: number) {
    return n > 0 ? 'yes' : null;
}
export function multiply(a: number, b: number) {
    return a * b;
}
export function approxPi() {
    return 3.14159;
}
