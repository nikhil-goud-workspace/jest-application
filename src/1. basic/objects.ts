export class User {
    constructor(public id: number, public name: string) { }
}

export function makeUser() {
    return {
        id: 1,
        name: 'Ava',
        meta: {
            roles: ['admin', 'editor'],
            flags: {
                active: true
            }
        }
    };
}

export function callAfter(ms: number, cb: () => void) {
    setTimeout(cb, ms);
}
