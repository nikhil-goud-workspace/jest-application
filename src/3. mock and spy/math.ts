
export const multiply = (a: number, b: number) => a * b;

export const sum = (a: number, b: number) => a + b;

// export var a = 1;