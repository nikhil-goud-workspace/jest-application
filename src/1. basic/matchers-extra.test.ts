import { add } from './math';
import { makeUser, User, callAfter } from './objects';
import { failIfNegative } from './errors';

describe('equality flavors', () => {
    test('toStrictEqual vs toEqual (class instance vs plain object)', () => {
        const u = new User(1, 'Ava');

        // toEqual ignores prototype and compares enumerable props -> passes
        expect(u).toEqual({ id: 1, name: 'Ava' });

        // toStrictEqual ALSO checks prototype -> fails if compared to plain object
        expect(u).not.toStrictEqual({ id: 1, name: 'Ava' });

        // Same prototype => passes
        expect(u).toStrictEqual(new User(1, 'Ava'));
    });


    test('toMatchObject for partial object match', () => {
        const obj = makeUser();
        expect(obj).toMatchObject({ name: 'Ava', meta: { flags: { active: true } } });
    });

    test('toContainEqual for deep array item equality', () => {
        const list = [{ a: 1 }, { b: 2 }];
        expect(list).toContainEqual({ a: 1 });
    });

    test('toHaveProperty with deep path + expected value', () => {
        const obj = makeUser();
        expect(obj).toHaveProperty('meta.roles.0', 'admin');
        expect(obj).toHaveProperty(['meta', 'flags', 'active'], true);
    });

    // {
    //     id: 1,
    //     name: 'Ava',
    //     meta: {
    //         roles: ['admin', 'editor'],
    //         flags: {
    //             active: true
    //         }
    //     }
    // };
    test('toBeDefined / toBeUndefined', () => {
        const obj: Record<string, unknown> = { a: 1, b: undefined };
        expect(obj.a).toBeDefined();
        expect(obj.b).toBeUndefined();
    });

    test('toHaveLength on string/array', () => {
        expect('Hello').toHaveLength(5);
        expect([1, 2, 3]).toHaveLength(3);
    });
});

describe('parameterized tests with test.each', () => {
    test.each([
        [1, 2, 3],
        [2, 3, 5],
        [10, -5, 5]
    ])('add(%i, %i) = %i', (a, b, expected) => {
        expect(add(a, b)).toBe(expected);
    });
});

describe('timers: jest.useFakeTimers', () => {
    test('advancing timers triggers callbacks', () => {
        jest.useFakeTimers();
        const spy = jest.fn();

        callAfter(1000, spy);
        expect(spy).not.toHaveBeenCalled();

        jest.advanceTimersByTime(1000);
        expect(spy).toHaveBeenCalledTimes(1);

        jest.useRealTimers();
    });
});

describe('snapshot: error text stays stable', () => {
    test('toThrowErrorMatchingSnapshot', () => {
        expect(() => failIfNegative(-1)).toThrowErrorMatchingSnapshot();
    });
});