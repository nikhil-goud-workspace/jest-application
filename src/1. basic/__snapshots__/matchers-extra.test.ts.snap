// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`snapshot: error text stays stable toThrowErrorMatchingSnapshot 1`] = `"Negative number!"`;
